08-optional-types.Rmd

```diff
@@ -0,0 +1,344 @@
+
+## Optional inputs
+The most common use case for optional variables are optional inputs for the user to provide some sort of file, for example, a reference genome or a metadata file. Sometimes, you will want optional inputs to fall back on a known file or value.
+
+A good place to use optional types in our workflow would be in the `bwa mem` task. `bwa mem` on default settings may attempt to use too many threads, which could cause this task to fail if it ran on weaker hardware such as a laptop. In order to make `bwa mem` run as expected on different backends, we can set `Int? bwa_mem_threads` as a workflow-level optional input. Making it optional means that users who do not want to think about threads -- perhaps they're using an HPC or other powerful backend -- do not have to worry about filling in that value. In the workflow body, when we call `bwa mem`, the value of `bwa_mem_threads` is passed to the `BwaMem` task.
+
+```
+  call BwaMem {
+    input:
+      input_fastq = sampleFastq,
+      ref_fasta = ref_fasta,
+      ref_fasta_index = ref_fasta_index,
+      ref_dict = ref_dict,
+      ref_amb = ref_amb,
+      ref_ann = ref_ann,
+      ref_bwt = ref_bwt,
+      ref_pac = ref_pac,
+      ref_sa = ref_sa
+      ref_sa = ref_sa,
+      threads = bwa_mem_threads
+  }
+```
+
+But what happens if the workflow-level variable `bwa_mem_threads` isn't defined? The task `BwaMem` will then get no value for the task-level variable `threads`. This will cause an error at runtime.
+
+One way to prevent this from happening is with WDL built-in function `select_first()`, which takes in an array of values. In the workflow body, we can use this plus a fallback value (or another variable) to ensure that the `BwaMem` task always gets a valid integer for the `threads` argument.
+
+```
+  call BwaMem {
+    input:
+      input_fastq = sampleFastq,
+      ref_fasta = ref_fasta,
+      ref_fasta_index = ref_fasta_index,
+      ref_dict = ref_dict,
+      ref_amb = ref_amb,
+      ref_ann = ref_ann,
+      ref_bwt = ref_bwt,
+      ref_pac = ref_pac,
+      ref_sa = ref_sa
+      ref_sa = ref_sa,
+      threads = select_first([bwa_mem_threads, 16])
+  }
+```
+
+This is a perfectly valid way of doing things, but the task itself still relies on that variable being defined. In the context of our workflow, that doesn't matter, but keep in mind it is a common practice to reuse code for multiple scripts -- whether that is directly importing other WDL tasks using WDL's importing feature, or simply copy-pasting boilerplate code. As such, it's more common to make the task itself handle this by providing a fallback value, like so:
```

Comment:
I'm a little confused: if we don't define `Int? bwa_mem_threads` at the workflow level, then `select_first()` will assign 16.  I'm agreeing that in the context of our workflow it doesn't matter, but I'm having trouble seeing the difference between ` threads = select_first([bwa_mem_threads, 16])` vs. `Int threads = 16`.

Reply:
Let me try to clarify the advice I want this chapter to get across...

This works, and is typical. By defining a fallback within the task, this makes the task easier to use by workflow(s). In this context, we're basically saying that no matter how you are running bwa_mem, it's safe to assume that 16 is a nice fallback value to use. In other words, how the number of threads we give bwa mem threads doesn't really care about the context of the workflow. bwa mem is used for a ton of things, but for most of those things, 16 threads is a pretty good way to go about it.

```
workflow whatever {
  inputs { 
    Int? bwa_mem_threads 
  }
  call task bwa_mem { 
    input: threads = bwa_mem_threads 
  }
}
task bwa_mem {
  inputs {
    Int threads = 16  # required variable with fallback
  }
  [...]
}
```

The most common alternative is falling back within the workflow, but it makes the bwa_mem task less "portable."  Any workflow that imports the bwa_mem task would need to use select_first() to manually define a fallback for `threads` which is one more detail to account for.

```
workflow whatever {
  inputs { 
    Int? bwa_mem_threads 
  }
  call task bwa_mem { 
    input: threads = select_first([bwa_mem_threads, 16])
  }
}
task bwa_mem {
  inputs {
    Int threads  # required variable with no fallback
  }
  [...]
}
```

This does not work, because `bwa_mem_threads` isn't defined in the scope of the task.
```
workflow whatever {
  inputs { 
    Int? bwa_mem_threads 
  }
  call task bwa_mem { 
    input: threads = bwa_mem_threads 
  }
}
task bwa_mem {
  inputs {
    Int threads = select_first([bwa_mem_threads, 16])  # borked
  }
  [...]
}
```

I'll take a look at the text and see if I can make the reasoning a bit clearer. But as you pointed out, there's a lot of scenarios where you have multiple options, so I don't want to go too deep down a rabbithole on something that isn't essential.

Reply:
Gotcha that makes sense to me. Portability seems key here.

Reply:
Reworded a few things -- take a look and let me know what you think.